File: web/src/lib/platforms.ts
import { apiClient } from './api'

export interface Platform {
  id: string
  slug: string           // unique identifier, e.g. "youtube"
  name: string           // display name, e.g. "YouTube"
  url: string            // base URL, e.g. "https://youtube.com"
  is_active: boolean     // whether the platform is enabled
  content_types: string[] // ["single", "playlist", "vod"]
  requires_login: boolean // does this platform need auth?
  supports_live: boolean  // can handle live streams?
  description: string | null
  limitations: string | null
  logo_url: string | null
  sort_order: number
  created_at: string | null
  updated_at: string | null
}

export interface CreatePlatformRequest {
  slug: string
  name: string
  url: string
  is_active?: boolean           // default true
  content_types?: string[]      // default ["single"]
  requires_login?: boolean      // default false
  supports_live?: boolean       // default false
  description?: string | null
  limitations?: string | null
  logo_url?: string | null
  sort_order?: number           // default 0
}

export type UpdatePlatformRequest = Partial<CreatePlatformRequest>

export interface PlatformListResponse {
  platforms: Platform[]
  total: number
}

/**
 * Fetch media platforms from the internal API.
 */
export async function getPlatforms(activeOnly: boolean = false): Promise<PlatformListResponse> {
  const endpoint = activeOnly ? '/platforms?active_only=true' : '/platforms'
  const res = await apiClient<{ status: string; data: PlatformListResponse }>(endpoint)
  return res.data
}

/**
 * Create a new supported media platform.
 */
export async function createPlatform(data: CreatePlatformRequest): Promise<Platform> {
  const res = await apiClient<{ status: string; data: Platform }>('/platforms', {
    method: 'POST',
    body: JSON.stringify(data),
  })
  return res.data
}

/**
 * Update configuration for an existing platform.
 */
export async function updatePlatform(id: string, data: UpdatePlatformRequest): Promise<Platform> {
  const res = await apiClient<{ status: string; data: Platform }>(`/platforms/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(data),
  })
  return res.data
}

/**
 * Permanently remove a platform from the catalog.
 */
export async function deletePlatform(id: string): Promise<void> {
  await apiClient<{ status: string; message: string }>(`/platforms/${id}`, {
    method: 'DELETE',
  })
}

File: backend/src/internal/schema/platforms.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

# ── Platform CRUD schemas ─────────────────────────────────────────────────────
# Request/response bodies for platform management endpoints.
# Content types are validated against the fixed set: single, playlist, vod.

ContentType = Literal["single", "playlist", "vod"]


class CreatePlatformRequest(BaseModel):
    """Request to create a new supported platform."""
    slug: str
    name: str
    url: str
    is_active: bool = True
    content_types: list[ContentType] = ["single"]
    requires_login: bool = False
    supports_live: bool = False
    description: str | None = None
    limitations: str | None = None
    logo_url: str | None = None
    sort_order: int = 0


class UpdatePlatformRequest(BaseModel):
    """Request to update a platform — all fields optional (partial update)."""
    slug: str | None = None
    name: str | None = None
    url: str | None = None
    is_active: bool | None = None
    content_types: list[ContentType] | None = None
    requires_login: bool | None = None
    supports_live: bool | None = None
    description: str | None = None
    limitations: str | None = None
    logo_url: str | None = None
    sort_order: int | None = None


class PlatformResponse(BaseModel):
    """Platform response — returned by all platform endpoints."""
    model_config = {"from_attributes": True}

    id: uuid.UUID
    slug: str
    name: str
    url: str
    is_active: bool
    content_types: list
    requires_login: bool
    supports_live: bool
    description: str | None = None
    limitations: str | None = None
    logo_url: str | None = None
    sort_order: int
    created_at: datetime | None = None
    updated_at: datetime | None = None


class PlatformListResponse(BaseModel):
    platforms: list[PlatformResponse]
    total: int
